mozetl/taar/taar_ensemble.py

```diff
@@ -0,0 +1,525 @@
+            "scalar_parent_browser_engagement_total_uri_count_sum AS total_uri",
+            "scalar_parent_browser_engagement_unique_domains_count_max AS unique_tlds",
+        )
+        .where("rn = 1")
+        .drop("rn")
+    )
+    return df
+
+
+def get_addons_per_client(users_df, minimum_addons_count):
+    """ Extracts a DataFrame that contains one row
+    for each client along with the list of active add-on GUIDs.
+    """
+
+    def is_valid_addon(addon):
+        return not (
+            addon.is_system
+            or addon.app_disabled
+            or addon.type != "extension"
+            or addon.user_disabled
+            or addon.foreign_install
+            or addon.install_day is None
+        )
+
+    # may need addiitonal whitelisting to remove shield addons
+
+    def get_valid_addon_ids(addons):
+        sorted_addons = sorted(
+            [(a.addon_id, a.install_day) for a in addons if is_valid_addon(a)],
+            key=lambda addon_tuple: addon_tuple[1],
+        )
+        return [addon_id for (addon_id, install_day) in sorted_addons]
+
+    get_valid_addon_ids_udf = udf(get_valid_addon_ids, ArrayType(StringType()))
+
+    # Create an add-ons dataset un-nesting the add-on map from each
+    # user to a list of add-on GUIDs. Also filter undesired add-ons.
+    return users_df.select(
+        "client_id", get_valid_addon_ids_udf("active_addons").alias("addon_ids")
+    ).filter(size("addon_ids") > minimum_addons_count)
+
+
+def safe_get_int(row, fieldname, default, factor=None):
+    tmp = getattr(row, fieldname, default)
+    if tmp is None:
+        return 0
+    try:
+        if factor is not None:
+            tmp *= factor
+        tmp = int(tmp)
+    except Exception:
+        return 0
+    return tmp
+
+
+def safe_get_str(row, fieldname):
+    tmp = getattr(row, fieldname, "")
+    if tmp is None:
+        return ""
+    return str(tmp)
+
+
+def row_to_json(row):
+    jdata = {}
+
+    # This is not entirely obvious.  All of our row data from raw telemetry uses *real*
+    # client_ids.   The production TAAR system only uses hashed telemetry client IDs.
+    # That said - we don't need to hash because we are only concerned
+    # with GUID recommendations here for the purposes of training
+    jdata["client_id"] = row.client_id
+
+    # Note the inconsistent naming of the addon ID field
+    jdata["installed_addons"] = row.addon_ids
+    jdata["bookmark_count"] = safe_get_int(row, "bookmark_count", 0)
+    jdata["tab_open_count"] = safe_get_int(row, "tab_open_count", 0)
+    jdata["total_uri"] = safe_get_int(row, "total_uri", 0)
+    jdata["subsession_length"] = safe_get_int(row, "subsession_length", 0, 3600)
+    jdata["unique_tlds"] = safe_get_int(row, "unique_tlds", 0)
+    jdata["geo_city"] = safe_get_str(row, "geo_city")
+    jdata["locale"] = safe_get_str(row, "locale")
+    jdata["os"] = safe_get_str(row, "os")
+
+    return jdata
+
+
+def reload_configuration():
```

Comment:
Docstring here would be nice. What's the reason for needing to dynamically reload the configuration?

Reply:
Added comment in 39d4b4d .

This is really a side effect of using python-decouple with TAAR.  The s3config module expects environment variables set before loading or else it uses empty string defaults.   When spark worker nodes are running - they don't have the enviroment - that's seems to be only sent to the spark driver node.